File: backend/app/detection/upi.py
from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import parse_qs, unquote

from app.detection.lexicon import load_json_config
from app.schemas.contracts import UpiFlag

logger = logging.getLogger(__name__)

_UPI_LINK_RE = re.compile(
    r"(?P<app>upi|phonepe|tez|paytmmp|gpay|bhim)://(?P<action>[a-z]+)\?(?P<query>\S*)",
    re.IGNORECASE,
)
# A UPI handle never contains a dot, which is what separates "raju@okaxis" from
# an email address like "help@sbi.co.in".
_VPA_RE = re.compile(r"(?<![\w.\-])[\w.\-]{2,64}@[a-z]{2,32}(?![\w.\-])", re.IGNORECASE)

_TERM_KEYS = ("en", "hi", "gu", "translit")
_ASCII_ONLY = re.compile(r"^[\x00-\x7f]+$")

_WEIGHTS = {
    "collect_request_disguised": 1.00,
    "pin_to_receive_money": 1.00,
    "unknown_vpa_payment": 0.50,
    "upi_link_present": 0.15,
}

_COLLECT_ACTIONS = frozenset({"collect", "mandate", "requestpay"})


def _compile_group(group: dict[str, list[str]] | None) -> re.Pattern[str] | None:
    """Compile one context group's term lists and raw regexes into a matcher."""
    if not group:
        return None
    patterns: list[str] = []
    for key in _TERM_KEYS:
        for term in group.get(key, []):
            body = r"\s+".join(re.escape(word) for word in term.split())
            patterns.append(
                rf"(?<!\w){body}(?:s|ed)?(?!\w)" if _ASCII_ONLY.match(term) else rf"(?<!\w){body}"
            )
    for raw in group.get("patterns", []):
        try:
            re.compile(raw)
        except re.error as exc:
            logger.warning("upi context pattern invalid pattern=%r error=%s", raw, exc)
            continue
        patterns.append(raw)
    if not patterns:
        return None
    try:
        return re.compile("|".join(patterns), re.IGNORECASE | re.UNICODE)
    except re.error as exc:
        logger.warning("upi context group failed to compile error=%s", exc)
        return None


@dataclass(frozen=True, slots=True)
class UpiIntent:
    raw: str
    action: str
    vpa: str | None
    payee_name: str | None
    amount: str | None
    note: str | None

    @property
    def is_collect(self) -> bool:
        return self.action.lower() in _COLLECT_ACTIONS


def _first(params: dict[str, list[str]], key: str) -> str | None:
    values = params.get(key)
    return unquote(values[0]) if values else None


def parse_upi_intents(text: str) -> list[UpiIntent]:
    intents: list[UpiIntent] = []
    for match in _UPI_LINK_RE.finditer(text):
        params = parse_qs(match.group("query"), keep_blank_values=False)
        intents.append(
            UpiIntent(
                raw=match.group(0),
                action=match.group("action"),
                vpa=_first(params, "pa"),
                payee_name=_first(params, "pn"),
                amount=_first(params, "am"),
                note=_first(params, "tn"),
            )
        )
    return intents


class UpiAnalyzer:
    """Detects the UPI traps that cost first-time users the most money.

    The central one: a UPI PIN authorises money leaving your account. Any message
    that pairs a promise of incoming money with a PIN entry or an approval prompt
    is inverting that fact, and is treated as a top-weight signal.
    """

    def __init__(self, config: dict) -> None:
        self.receive_claim = _compile_group(config.get("receive_claims"))
        self.pin_instruction = _compile_group(config.get("pin_instructions"))
        self.collect_instruction = _compile_group(config.get("collect_instructions"))
        self.known_handles: frozenset[str] = frozenset(
            handle.lower().lstrip("@") for handle in config.get("known_handles", [])
        )

    @classmethod
    def from_file(cls, path: Path) -> UpiAnalyzer:
        analyzer = cls(load_json_config(path))
        logger.info("upi context loaded handles=%d", len(analyzer.known_handles))
        return analyzer

    def analyze(self, text: str) -> list[UpiFlag]:
        intents = parse_upi_intents(text)
        mentioned_vpa = next(iter(_VPA_RE.findall(text)), None)
        claims_incoming = self._matches(self.receive_claim, text)
        asks_for_pin = self._matches(self.pin_instruction, text)
        asks_to_approve = self._matches(self.collect_instruction, text)
        has_collect_intent = any(intent.is_collect for intent in intents)

        flags: list[UpiFlag] = []

        if claims_incoming and asks_for_pin:
            flags.append(
                self._flag(
                    "pin_to_receive_money",
                    "The message promises you money and then asks for your UPI PIN. "
                    "A PIN only ever sends money out of your account.",
                    intents,
                    mentioned_vpa,
                )
            )

        if claims_incoming and (has_collect_intent or asks_to_approve):
            flags.append(
                self._flag(
                    "collect_request_disguised",
                    "This is a request to take money from your account, dressed up as money coming in. "
                    "Approving it pays them.",
                    intents,
                    mentioned_vpa,
                )
            )

        if not flags:
            for intent in intents:
                if self._is_unknown_payee(intent):
                    flags.append(
                        self._flag(
                            "unknown_vpa_payment",
                            "The link pays a UPI address that is not a recognised payment provider.",
                            [intent],
                        )
                    )
                    break

        if not flags and intents:
            flags.append(
                self._flag(
                    "upi_link_present",
                    "The message contains a payment link. Check the name and the "
                    "direction in your app before approving.",
                    intents,
                )
            )

        return flags

    def _is_unknown_payee(self, intent: UpiIntent) -> bool:
        if intent.is_collect:
            return True
        if not intent.vpa or "@" not in intent.vpa:
            return False
        handle = intent.vpa.rsplit("@", 1)[1].lower()
        return bool(self.known_handles) and handle not in self.known_handles

    @staticmethod
    def _matches(pattern: re.Pattern[str] | None, text: str) -> bool:
        return pattern is not None and pattern.search(text) is not None

    @staticmethod
    def _flag(
        reason: str, detail: str, intents: list[UpiIntent], mentioned_vpa: str | None = None
    ) -> UpiFlag:
        intent = intents[0] if intents else None
        return UpiFlag(
            reason=reason,
            detail=detail,
            weight=_WEIGHTS[reason],
            vpa=(intent.vpa if intent else None) or mentioned_vpa,
            amount=intent.amount if intent else None,
        )

File: frontend/src/lib/engine/upi.ts
/**
 * Port of backend/app/detection/upi.py.
 *
 * The central trap: a UPI PIN authorises money leaving your account, so any
 * message pairing a promise of incoming money with a PIN entry or an approval
 * prompt is a top-weight signal. Weights and reason codes match the server.
 */

import type { EngineUpiFlag as UpiFlag, UpiConfig, UpiGroup } from "@/lib/engine/types";

const UPI_LINK_RE = /(upi|phonepe|tez|paytmmp|gpay|bhim):\/\/([a-z]+)\?(\S*)/gi;
// A UPI handle never contains a dot, which separates it from an email address.
const VPA_RE = /(?<![\w.-])[\w.-]{2,64}@[a-z]{2,32}(?![\w.-])/gi;
const TERM_KEYS = ["en", "hi", "gu", "translit"] as const;
const ASCII_ONLY = /^[\x00-\x7f]+$/;
const WB = "[\\p{L}\\p{N}_]";
const COLLECT_ACTIONS = new Set(["collect", "mandate", "requestpay"]);

const WEIGHTS: Record<string, number> = {
  collect_request_disguised: 1.0,
  pin_to_receive_money: 1.0,
  unknown_vpa_payment: 0.5,
  upi_link_present: 0.15,
};

function escapeRegExp(word: string): string {
  return word.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function compileGroup(group: UpiGroup | undefined): RegExp | null {
  if (!group) return null;
  const patterns: string[] = [];
  for (const key of TERM_KEYS) {
    for (const term of group[key] ?? []) {
      const body = term.split(/\s+/).filter(Boolean).map(escapeRegExp).join("\\s+");
      patterns.push(ASCII_ONLY.test(term) ? `(?<!${WB})${body}(?:s|ed)?(?!${WB})` : `(?<!${WB})${body}`);
    }
  }
  for (const raw of group.patterns ?? []) patterns.push(raw);
  if (patterns.length === 0) return null;
  try {
    return new RegExp(patterns.join("|"), "giu");
  } catch {
    try {
      return new RegExp(patterns.join("|"), "gi");
    } catch {
      return null;
    }
  }
}

function decode(value: string): string {
  try {
    return decodeURIComponent(value.replace(/\+/g, " "));
  } catch {
    return value;
  }
}

interface UpiIntent {
  raw: string;
  action: string;
  vpa: string | null;
  amount: string | null;
}

function parseIntents(text: string): UpiIntent[] {
  const intents: UpiIntent[] = [];
  UPI_LINK_RE.lastIndex = 0;
  for (const m of text.matchAll(UPI_LINK_RE)) {
    const params = new Map<string, string>();
    for (const pair of m[3]!.split("&")) {
      if (!pair) continue;
      const eq = pair.indexOf("=");
      const key = eq === -1 ? pair : pair.slice(0, eq);
      const val = eq === -1 ? "" : pair.slice(eq + 1);
      if (val && !params.has(key)) params.set(key, decode(val));
    }
    intents.push({
      raw: m[0],
      action: m[2]!,
      vpa: params.get("pa") ?? null,
      amount: params.get("am") ?? null,
    });
  }
  return intents;
}

function isCollect(intent: UpiIntent): boolean {
  return COLLECT_ACTIONS.has(intent.action.toLowerCase());
}

function matches(pattern: RegExp | null, text: string): boolean {
  if (!pattern) return false;
  pattern.lastIndex = 0;
  return pattern.test(text);
}

export class UpiAnalyzer {
  private readonly receiveClaim: RegExp | null;
  private readonly pinInstruction: RegExp | null;
  private readonly collectInstruction: RegExp | null;
  private readonly knownHandles: Set<string>;

  constructor(config: UpiConfig) {
    this.receiveClaim = compileGroup(config.receive_claims);
    this.pinInstruction = compileGroup(config.pin_instructions);
    this.collectInstruction = compileGroup(config.collect_instructions);
    this.knownHandles = new Set((config.known_handles ?? []).map((h) => h.toLowerCase().replace(/^@/, "")));
  }

  analyze(text: string): UpiFlag[] {
    const intents = parseIntents(text);
    VPA_RE.lastIndex = 0;
    const mentionedVpa = text.match(VPA_RE)?.[0] ?? null;
    const claimsIncoming = matches(this.receiveClaim, text);
    const asksForPin = matches(this.pinInstruction, text);
    const asksToApprove = matches(this.collectInstruction, text);
    const hasCollectIntent = intents.some(isCollect);

    const flags: UpiFlag[] = [];

    if (claimsIncoming && asksForPin) {
      flags.push(
        flagFor(
          "pin_to_receive_money",
          "The message promises you money and then asks for your UPI PIN. A PIN only ever sends money out of your account.",
          intents,
          mentionedVpa,
        ),
      );
    }

    if (claimsIncoming && (hasCollectIntent || asksToApprove)) {
      flags.push(
        flagFor(
          "collect_request_disguised",
          "This is a request to take money from your account, dressed up as money coming in. Approving it pays them.",
          intents,
          mentionedVpa,
        ),
      );
    }

    if (flags.length === 0) {
      for (const intent of intents) {
        if (this.isUnknownPayee(intent)) {
          flags.push(
            flagFor("unknown_vpa_payment", "The link pays a UPI address that is not a recognised payment provider.", [intent]),
          );
          break;
        }
      }
    }

    if (flags.length === 0 && intents.length > 0) {
      flags.push(
        flagFor(
          "upi_link_present",
          "The message contains a payment link. Check the name and the direction in your app before approving.",
          intents,
        ),
      );
    }

    return flags;
  }

  private isUnknownPayee(intent: UpiIntent): boolean {
    if (isCollect(intent)) return true;
    if (!intent.vpa || !intent.vpa.includes("@")) return false;
    const handle = intent.vpa.split("@").pop()!.toLowerCase(); // vpa contains "@"
    return this.knownHandles.size > 0 && !this.knownHandles.has(handle);
  }
}

function flagFor(reason: string, detail: string, intents: UpiIntent[], mentionedVpa: string | null = null): UpiFlag {
  const intent = intents[0] ?? null;
  return {
    reason,
    detail,
    weight: WEIGHTS[reason]!,
    vpa: (intent?.vpa ?? null) || mentionedVpa,
    amount: intent?.amount ?? null,
  };
}
